### ISSUE TITLE
sql: "opening" join filters incorrect in some cases

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#19640]
We have code that converts a post-join filter like `x > 10` into `a.x > 10 AND b.x > 10`. This is not quite correct for left or right outer joins, where one of `a.x, b.x` can be NULL.
```sql
root@:26257/test> show create table a;
+-------+----------------------------------------+
| Table |              CreateTable               |
+-------+----------------------------------------+
| a     | CREATE TABLE a (␤                      |
|       |     x INT NULL,␤                       |
|       |     y INT NULL,␤                       |
|       |     u INT NULL,␤                       |
|       |     FAMILY "primary" (x, y, u, rowid)␤ |
|       | )                                      |
+-------+----------------------------------------+
(1 row)

root@:26257/test> show create table b;
+-------+----------------------------------------+
| Table |              CreateTable               |
+-------+----------------------------------------+
| b     | CREATE TABLE b (␤                      |
|       |     x INT NULL,␤                       |
|       |     y INT NULL,␤                       |
|       |     v INT NULL,␤                       |
|       |     FAMILY "primary" (x, y, v, rowid)␤ |
|       | )                                      |
+-------+----------------------------------------+
(1 row)

root@:26257/test> select *,a.x,b.x from a left outer join b USING (x,y);
+---+---+---+------+---+------+
| x | y | u |  v   | x |  x   |
+---+---+---+------+---+------+
| 1 | 1 | 1 | NULL | 1 | NULL |
| 2 | 2 | 2 |    2 | 2 |    2 |
+---+---+---+------+---+------+
(2 rows)

root@:26257/test> select *,a.x,b.x from a left outer join b USING (x,y) where x is not null;
+---+---+---+---+---+---+
| x | y | u | v | x | x |
+---+---+---+---+---+---+
| 2 | 2 | 2 | 2 | 2 | 2 |
+---+---+---+---+---+---+
(1 row)

root@:26257/test> select *,a.x,b.x from a left outer join b USING (x,y) where x > 0;
+---+---+---+---+---+---+
| x | y | u | v | x | x |
+---+---+---+---+---+---+
| 2 | 2 | 2 | 2 | 2 | 2 |
+---+---+---+---+---+---+
(1 row)
```

**Comment by RaduBerinde** [comment]
CC @knz

**Comment by knz** [comment]
yeah, I messed up when I did the outer join stuff. My bad. Thanks for spotting it.

**Comment by knz** [comment]
I think this has been fixed in  #19683?
